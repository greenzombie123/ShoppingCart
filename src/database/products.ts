export type ProductCategory = "Jewelry" | "Men's Clothing" | "Women's Clothing" | "Electronics"

export type Product = {
	id:number,
	name:string,
	price:number,
	rating:number,
	stars:1|2|3|4|5,
	likes:number,
	styles:Style[]
	category:ProductCategory
}

export type Style = {
	description:string,
	picture:URL,
	isCurrentStyle:boolean, 
}

export type CartItem = {
	product:Product,
	id:number,
	quantity:number,
	style:string
}

export type Cart = CartItem[]