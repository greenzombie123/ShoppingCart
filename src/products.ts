// import fs from "fs";

export type ProductCategory =
  | "Jewelry"
  | "Men's Clothing"
  | "Women's Clothing"
  | "Electronics";

export type Product = {
  id: number | string;
  name: string;
  price: number;
  ratings: number;
  stars: 1 | 2 | 3 | 4 | 5;
  likes: number;
  styles: Style[];
  category: ProductCategory;
};

export type Style = {
  description: string;
  picture: string;
  isCurrentStyle: boolean;
};

export type CartItem = {
  name:string,
  price: number;
  id: number;
  quantity: number;
  style: string | undefined,
  picture:string
};

export type Cart = CartItem[];

// const thing: Product[] = [
//   {
//     id: 1,
//     name: "polo shirt",
//     price: 2314,
//     ratings: 334,
//     stars: 5,
//     likes: 380,
//     category: "Men's Clothing",
//     styles: [
//       {
//         description: "black",
//         picture: "/images/poloshirtblack.webp",
//         isCurrentStyle: true,
//       },
//       {
//         description: "blue",
//         picture: "/images/bluepoloshirt.jpg",
//         isCurrentStyle: false,
//       },
//     ],
//   },
//   {
//     id: 1,
//     name: "shirt",
//     price: 1299,
//     ratings: 234,
//     stars: 4,
//     likes: 1000,
//     category: "Men's Clothing",
//     styles: [
//       {
//         description: "white",
//         picture: "src/assets/bluepoloshirt.jpg",
//         isCurrentStyle: true,
//       },
//     ],
//   },
// ];

// const doShit = async () => {
//   const g = await fetch(`http://localhost:3000/test`, {
//     method: "PATCH",
//     headers: { "Content-Type": "application/json" },
//     body:JSON.stringify({test:[]})
//   }).catch(e=>console.log(e))



//   console.log(g)

// };

// doShit()
