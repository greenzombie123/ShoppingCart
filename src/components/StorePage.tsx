import { Link } from "react-router-dom";
import style from "./StorePage.module.css";
import { Product } from "../products";

type StoreItemProps = {
  product:Product
}

export const StoreItem = ({product}:StoreItemProps) => {

  return (
    <div className={style.itemCard}>
      <Link to={`/product/${product.id}`}>
        <div className={style.imgContainer}>
          <img src={product.styles[0].picture} alt={product.name} />
          <div className={style.bestSeller}>BestSeller</div>
          <div className={style.heartContainer}>
            <div className={style.heart}></div>
            <p className={style.number}>2</p>
          </div>
        </div>
        <div className={style.itemInfo}>
          <div className={style.styles}></div>
          <div className={style.name}>{product.name}</div>
          <div className={style.stars}>
            <div className={style.stars_container}>
              <div className={style.star + " " + style.on}></div>
              <div className={style.star + " " + style.on}></div>
              <div className={style.star + " " + style.on}></div>
              <div className={style.star}></div>
              <div className={style.star}></div>
            </div>
            <div className={style.ratings}>({product.ratings})</div>
          </div>
          <div className={style.price}>${product.price}</div>
        </div>
      </Link>
    </div>
  );
};

const StorePage = () => {
  return <div className={style.items} data-testid="store_page">
    
  </div>;
};

export default StorePage;
