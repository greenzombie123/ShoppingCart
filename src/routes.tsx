import { RouteObject } from "react-router-dom";
import App from "./App";
import { addToCart, getCart, getRandomProducts, getStoreItems } from "./Loaders";
import Carousel from "./components/Carousel";
import StorePage from "./components/StorePage";
import ShoppingProduct from "./components/ShoppingProduct";
import ShoppingCart from "./components/ShoppingCart";

const routes: RouteObject[] = [
  {
    path: "/",
    element: <App />,
    loader: getCart,
    children: [
      {
        loader:getRandomProducts,
        element: <Carousel />,
        index:true,
      },
      {
        loader:getStoreItems,
        element:<StorePage/>,
        path:"store/:category"
      },
      {
        action:addToCart,
        element:<ShoppingProduct/>,
        path:"product/:id"
      },
      {
        action:getCart,
        element:<ShoppingCart/>,
        path:"mycart"
      }
    ],
  },
];

export default routes;
