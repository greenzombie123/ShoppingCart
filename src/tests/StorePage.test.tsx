import { afterAll, beforeAll, describe, expect, it, Mock, vi } from "vitest";
import { Cart, Product } from "../products";
import {
  createMemoryRouter,
  RouteObject,
  RouterProvider,
} from "react-router-dom";
import StorePage, { StarContainer } from "../components/StorePage";
import { render, screen, waitFor, within } from "@testing-library/react";
import routes from "../routes";
import userEvent from "@testing-library/user-event";
import storePageStyle from "../components/StorePage.module.css";

const products: Product[] = [
  {
    id: 1,
    name: "polo shirt",
    price: 2314,
    ratings: 334,
    stars: 2,
    likes: 380,
    category: "Men's Clothing",
    styles: [
      {
        description: "black",
        picture: "/images/poloshirtblack.webp",
        isCurrentStyle: true,
      },
      {
        description: "blue",
        picture: "/images/bluepoloshirt.jpg",
        isCurrentStyle: false,
      },
    ],
  },
  {
    id: 2,
    name: "polo shirta",
    price: 2314,
    ratings: 334,
    stars: 5,
    likes: 380,
    category: "Jewelry",
    styles: [
      {
        description: "black",
        picture: "/images/poloshirtblack.webp",
        isCurrentStyle: true,
      },
      {
        description: "blue",
        picture: "/images/bluepoloshirt.jpg",
        isCurrentStyle: false,
      },
    ],
  },
  {
    id: 3,
    name: "polo shirtb",
    price: 2314,
    ratings: 334,
    stars: 5,
    likes: 380,
    category: "Men's Clothing",
    styles: [
      {
        description: "black",
        picture: "/images/poloshirtblack.webp",
        isCurrentStyle: true,
      },
      {
        description: "blue",
        picture: "/images/bluepoloshirt.jpg",
        isCurrentStyle: false,
      },
    ],
  },
  {
    id: 4,
    name: "polo shirtc",
    price: 2314,
    ratings: 334,
    stars: 5,
    likes: 380,
    category: "Electronics",
    styles: [
      {
        description: "black",
        picture: "/images/poloshirtblack.webp",
        isCurrentStyle: true,
      },
      {
        description: "blue",
        picture: "/images/bluepoloshirt.jpg",
        isCurrentStyle: false,
      },
    ],
  },
  {
    id: 5,
    name: "polo shirtd",
    price: 2314,
    ratings: 334,
    stars: 5,
    likes: 380,
    category: "Women's Clothing",
    styles: [
      {
        description: "black",
        picture: "/images/poloshirtblack.webp",
        isCurrentStyle: true,
      },
      {
        description: "blue",
        picture: "/images/bluepoloshirt.jpg",
        isCurrentStyle: false,
      },
    ],
  },
];
const menClothingProducts = [products[0], products[2]];
const emptyCart: Cart = [];
const productsURL = "http://localhost:3000/products";
const cartURL = "http://localhost:3000/cart";
const menClothingURL = "http://localhost:3000/products?category=Men's Clothing";

beforeAll(() => {
  HTMLElement.prototype.scrollIntoView = vi.fn();

  vi.spyOn(global, "fetch").mockImplementation(
    vi.fn((url: string) => {
      const values =
        url === productsURL
          ? products
          : cartURL === url
          ? emptyCart
          : url === menClothingURL
          ? menClothingProducts
          : null;
      if (!values) throw new Error("Couldn't process url string");
      return Promise.resolve({
        json: () => Promise.resolve(values),
      });
    }) as Mock
  );
});

afterAll(() => {
  vi.resetAllMocks();
});

describe("StorePage", () => {
  it("renders on the screen", async () => {
    const route: RouteObject[] = [
      { path: "/", element: <StorePage />, loader: () => products },
    ];
    const router = createMemoryRouter(route);
    render(<RouterProvider router={router} />);

    const storePage = await waitFor(() => screen.getByTestId("store_page"));

    expect(storePage).toBeInTheDocument();
    expect(storePage).toMatchSnapshot();
  });

  it("render store page when men's clothing link is clicked", async () => {
    const router = createMemoryRouter(routes);
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    const link = await waitFor(() =>
      screen.getByRole("link", { name: "Men's Clothing" })
    );

    expect(link).toBeInTheDocument();

    await user.click(link);

    const storePage = await waitFor(() => screen.getByTestId("store_page"));

    expect(storePage).toBeInTheDocument();
  });

  it("render store items when men's clothing link is clicked", async () => {
    const router = createMemoryRouter(routes);
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    const link = await waitFor(() =>
      screen.getByRole("link", { name: "Men's Clothing" })
    );

    await user.click(link);

    const storePage = await waitFor(() => screen.getByTestId("store_page"));

    // Use within to find DOM elements within DOM elements
    const storeItems = await waitFor(() =>
      within(storePage).getAllByAltText(/^polo shirt/i)
    );

    expect(storeItems[0]).toBeInTheDocument();
    expect(storeItems[1]).toBeInTheDocument();
  });

  it(" renders all 5 stars in star container", async () => {
    render(<StarContainer stars={0} />);

    const stars = await waitFor(() =>
      screen.getAllByRole("img", { name: "star" })
    );
    // console.log(await screen.getByRole("img", {name:"star"}).classList)
    expect(stars.length).toBe(5);
  });

  it(" renders 4 stars that are on", async () => {
    render(<StarContainer stars={4} />);

    const stars = await waitFor(() =>
      screen
        .getAllByRole("img", { name: "star" })
        .filter((star) => star.classList.contains(storePageStyle.on))
    );
    // console.log(await screen.getByRole("img", {name:"star"}).classList)
    expect(stars.length).toBe(4);
  });

  it(" renders star container", async () => {
    const route: RouteObject[] = [
      { path: "/", element: <StorePage />, loader: () => [products[0]] },
    ];
    const router = createMemoryRouter(route);
    render(<RouterProvider router={router} />);

    const stars = await waitFor(() =>
      screen.getAllByRole("img", { name: "star" })
    );
    
    expect(stars.length).toBe(5);
  });
});
