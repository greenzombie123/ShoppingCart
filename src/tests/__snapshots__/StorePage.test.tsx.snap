// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StorePage > renders on the screen 1`] = `
<div
  class="_items_079726"
  data-testid="store_page"
>
  <div
    class="_itemCard_079726"
  >
    <a
      data-discover="true"
      href="/product/1"
    >
      <div
        class="_imgContainer_079726"
      >
        <img
          alt="polo shirt"
          src="/images/poloshirtblack.webp"
        />
        <div
          class="_bestSeller_079726"
        >
          BestSeller
        </div>
        <div
          class="_heartContainer_079726"
        >
          <div
            class="_heart_079726"
          />
          <p
            class="_number_079726"
          >
            2
          </p>
        </div>
      </div>
      <div
        class="_itemInfo_079726"
      >
        <div
          class="_styles_079726"
        />
        <div
          class="_name_079726"
        >
          polo shirt
        </div>
        <div
          class="_stars_079726"
        >
          <div
            class="_stars_container_079726"
          >
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726"
              role="img"
            />
          </div>
          <div
            class="_ratings_079726"
          >
            (
            334
            )
          </div>
        </div>
        <div
          class="_price_079726"
        >
          $
          2314
        </div>
      </div>
    </a>
  </div>
  <div
    class="_itemCard_079726"
  >
    <a
      data-discover="true"
      href="/product/2"
    >
      <div
        class="_imgContainer_079726"
      >
        <img
          alt="polo shirta"
          src="/images/poloshirtblack.webp"
        />
        <div
          class="_bestSeller_079726"
        >
          BestSeller
        </div>
        <div
          class="_heartContainer_079726"
        >
          <div
            class="_heart_079726"
          />
          <p
            class="_number_079726"
          >
            2
          </p>
        </div>
      </div>
      <div
        class="_itemInfo_079726"
      >
        <div
          class="_styles_079726"
        />
        <div
          class="_name_079726"
        >
          polo shirta
        </div>
        <div
          class="_stars_079726"
        >
          <div
            class="_stars_container_079726"
          >
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
          </div>
          <div
            class="_ratings_079726"
          >
            (
            334
            )
          </div>
        </div>
        <div
          class="_price_079726"
        >
          $
          2314
        </div>
      </div>
    </a>
  </div>
  <div
    class="_itemCard_079726"
  >
    <a
      data-discover="true"
      href="/product/3"
    >
      <div
        class="_imgContainer_079726"
      >
        <img
          alt="polo shirtb"
          src="/images/poloshirtblack.webp"
        />
        <div
          class="_bestSeller_079726"
        >
          BestSeller
        </div>
        <div
          class="_heartContainer_079726"
        >
          <div
            class="_heart_079726"
          />
          <p
            class="_number_079726"
          >
            2
          </p>
        </div>
      </div>
      <div
        class="_itemInfo_079726"
      >
        <div
          class="_styles_079726"
        />
        <div
          class="_name_079726"
        >
          polo shirtb
        </div>
        <div
          class="_stars_079726"
        >
          <div
            class="_stars_container_079726"
          >
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
          </div>
          <div
            class="_ratings_079726"
          >
            (
            334
            )
          </div>
        </div>
        <div
          class="_price_079726"
        >
          $
          2314
        </div>
      </div>
    </a>
  </div>
  <div
    class="_itemCard_079726"
  >
    <a
      data-discover="true"
      href="/product/4"
    >
      <div
        class="_imgContainer_079726"
      >
        <img
          alt="polo shirtc"
          src="/images/poloshirtblack.webp"
        />
        <div
          class="_bestSeller_079726"
        >
          BestSeller
        </div>
        <div
          class="_heartContainer_079726"
        >
          <div
            class="_heart_079726"
          />
          <p
            class="_number_079726"
          >
            2
          </p>
        </div>
      </div>
      <div
        class="_itemInfo_079726"
      >
        <div
          class="_styles_079726"
        />
        <div
          class="_name_079726"
        >
          polo shirtc
        </div>
        <div
          class="_stars_079726"
        >
          <div
            class="_stars_container_079726"
          >
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
          </div>
          <div
            class="_ratings_079726"
          >
            (
            334
            )
          </div>
        </div>
        <div
          class="_price_079726"
        >
          $
          2314
        </div>
      </div>
    </a>
  </div>
  <div
    class="_itemCard_079726"
  >
    <a
      data-discover="true"
      href="/product/5"
    >
      <div
        class="_imgContainer_079726"
      >
        <img
          alt="polo shirtd"
          src="/images/poloshirtblack.webp"
        />
        <div
          class="_bestSeller_079726"
        >
          BestSeller
        </div>
        <div
          class="_heartContainer_079726"
        >
          <div
            class="_heart_079726"
          />
          <p
            class="_number_079726"
          >
            2
          </p>
        </div>
      </div>
      <div
        class="_itemInfo_079726"
      >
        <div
          class="_styles_079726"
        />
        <div
          class="_name_079726"
        >
          polo shirtd
        </div>
        <div
          class="_stars_079726"
        >
          <div
            class="_stars_container_079726"
          >
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
            <span
              aria-label="star"
              class="_star_079726 _on_079726"
              role="img"
            />
          </div>
          <div
            class="_ratings_079726"
          >
            (
            334
            )
          </div>
        </div>
        <div
          class="_price_079726"
        >
          $
          2314
        </div>
      </div>
    </a>
  </div>
</div>
`;
